---
weight: 30
sourceSHA: ebb22e5b509d0b31307726d79479c6c9accccd4df4baa7bcf1456aea72add281
---

# Обработка исключений в базе данных Mnesia RabbitMQ

## Общие сбои базы данных Mnesia

RabbitMQ использует базу данных Mnesia для хранения информации, такой как очереди, обмены и привязки. Основные причины сбоев Mnesia можно разделить на два типа:

- Проблемы с разрешениями в каталоге MNESIA\_BASE, когда у пользователя недостаточно прав на запись в этот каталог.
- Mnesia не может прочитать таблицу.

## Проблемы с разрешениями

Когда вы сталкиваетесь с проблемами разрешений в базе данных Mnesia, просто настройте права на запись для текущего пользователя в каталоге MNESIA\_BASE, чтобы решить проблему.

## Ошибка чтения таблицы Mnesia

Mnesia создает соответствующую схему базы данных на основе имени хоста машины. Поэтому, когда имя хоста меняется, Mnesia не может загрузить старую схему. Аналогично, если каталог rabbit\@hostname переименован, Mnesia также не сможет найти старые файлы базы данных, что заставит её создать новую папку rabbit\@hostname и начать базу данных заново.

Чтобы исправить ошибки чтения Mnesia, вы можете выполнить следующие шаги:

1. Сначала проверьте, были ли изменения в имени хоста узла или имени каталога rabbit\@hostname. Если изменения были, вы можете переименовать его, следуя соглашению “rabbit\@hostname”, что позволит вам увидеть старые файлы базы данных после переименования.

2. Если обновлений не обнаружено, в режиме кластера запуск может завершиться неудачно из-за неудачных соединений между репликами. Вы можете изучить проблемы связи между репликами на основе журналов.

3. Если вышеуказанные действия не восстанавливают данные, вы можете выбрать вариант резервного копирования конфигурации текущего кластера, развернуть новые узлы для присоединения к оригинальному кластеру и импортировать резервную конфигурацию.

## Другие сбои, связанные с базой данных Mnesia

Если RabbitMQ не может правильно сформировать новый кластер после перезагрузки, вы можете столкнуться с ошибками в журнале, подобными следующим:

```
[warning] <0.273.0> Ошибка при ожидании таблиц Mnesia: {timeout_waiting_for_tables,['rabbit@e2e-rabbitmq-server-2.e2e-rabbitmq-nodes.local-midautons','rabbit@e2e-rabbitmq-server-1.e2e-rabbitmq-nodes.local-midautons','rabbit@e2e-rabbitmq-server-0.e2e-rabbitmq-nodes.local-midautons'],[rabbit_durable_queue]}
 [info] <0.273.0> Ожидание таблиц Mnesia в течение 30000 мс, осталось 7 повторных попыток
```

Эта проблема возникает потому, что информация о узлах кластера хранится в базе данных Mnesia, и информация о состоянии не была должным образом очищена во время перезапуска RabbitMQ. Из-за последовательной политики запуска RabbitMQ для Pods это может привести к тому, что первоначально запущенные Pods будут бесконечно ожидать завершения запуска других Pods, создавая взаимную блокировку. Рекомендуемое решение — принудительное удаление состояния кластера и его перезапуск. Вы можете выполнить следующую команду в kubectl:

```
kubectl -n {имя пространства имен} exec {имя экземпляра} -- rabbitmqctl force_boot
```
